src/cascadia/TerminalCore/Terminal.cpp

```diff
@@ -1626,11 +1627,37 @@ void Terminal::PreviewText(std::wstring_view input)
         // pad it out
         preview.insert(originalSize, expectedLenTillEnd - originalSize, L' ');
     }
-    snippetPreview.text = til::visualize_nonspace_control_codes(preview);
+
     // Build our composition data
+    // The text is just the trimmed command, with the spaces at the end.
+    snippetPreview.text = til::visualize_nonspace_control_codes(preview);
+
+    // The attributes depend on the $profile:experimental.rainbowSuggestions setting:.
     const auto len = snippetPreview.text.size();
     snippetPreview.attributes.clear();
-    snippetPreview.attributes.emplace_back(len, previewAttrs);
+
+    if (_rainbowSuggestions)
+    {
+        // Let's do something fun.
+
+        // Use the actual text length for the number of steps, not including the
+        // trailing spaces.
+        const float increment = 1.0f / originalSize;

```

Comment:
Shouldn't this use `len` instead?

Reply:
In that case you shouldn't append 

But that confirms that `originalSize` can be less than `len`, right? The problem is that the total length of all `snippetPreview.attributes` should equal `snippetPreview.text.size()` (= `len`).

The current `Renderer` code iterates through the list of `.attributes` and grabs the chunk of text it refers to. This means that if your `.attributes` refer to 20 characters worth of text and the `.text` contains 30 characters, then it will simply only draw the first 20 characters. In other words, the trailing whitespaces will be ignored with this code.

To solve this, I think you need to simply append a final attribute that covers the trailing whitespaces.